source/extensions/filters/http/oauth/filter.h

```diff
@@ -0,0 +1,217 @@
+public:
+  virtual ~CookieValidator() = default;
+  virtual const std::string& username() const PURE;
+  virtual const std::string& token() const PURE;
+  virtual void setParams(const Http::RequestHeaderMap& headers, const std::string& secret) PURE;
+  virtual bool isValid() const PURE;
+};
+
+class OAuth2CookieValidator : public CookieValidator {
+public:
+  OAuth2CookieValidator() = default;
+  ~OAuth2CookieValidator() override = default;
+  const std::string& username() const override { return username_; }
+  const std::string& token() const override { return token_; }
+  void setParams(const Http::RequestHeaderMap& headers, const std::string& secret) override;
+  bool isValid() const override;
+  bool hmacIsValid() const;
+  bool timestampIsValid() const;
+
+private:
+  std::string username_;
+  std::string token_;
+  std::string expires_;
+  std::string hmac_;
+  std::vector<uint8_t> secret_;
+  absl::string_view host_;
+};
+
+/**
+ * The filter is the primary entrypoint for the OAuth workflow. Its responsibilities are to
+ * receive incoming requests and decide at what state of the OAuth workflow they are in. Logic
+ * beyond that is broken into component classes.
+ */
+class OAuth2Filter : public Http::PassThroughDecoderFilter, public OAuth2FilterCallbacks {
+public:
+  OAuth2Filter(OAuth2FilterConfigSharedPtr config, std::unique_ptr<OAuth2Client>&& oauth_client,
+               Stats::Scope& scope);
+  ~OAuth2Filter() override = default;
+
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers, bool) override;
+  void onGetAccessTokenSuccess(const std::string& access_code,
+                               const std::string& expires_in) override;
+  void onGetIdentitySuccess(const std::string& username) override;
+  // a catch-all function used for request failures. we don't retry, as a user can simply refresh
+  // the page in the case of a network blip.
+  void sendUnauthorizedResponse() override;
+
+  // Set the x-forwarded-user after successfully validating the client cookies.
+  static void setXForwardedOauthHeaders(Http::RequestHeaderMap& headers, const std::string& token,
+                                        const std::string& username);
+
+private:
+  friend class OAuth2Test;
+
+  using RequirementsMap = absl::flat_hash_map<absl::string_view, absl::string_view>;
+
+  std::shared_ptr<CookieValidator> validator_;
+
+  // wrap up some of these in a UserData struct or something...
+  std::string auth_code_{};
+  std::string access_token_{}; // TODO - see if we can avoid this being a member variable
+  std::string username_;
+  std::string new_expires_;
+  absl::string_view host_;
+  std::string user_agent_;
+  std::string state_{};
+  bool found_bearer_token_{false};
+  Http::RequestHeaderMap* request_headers_{nullptr};
+
+  std::unique_ptr<OAuth2Client> oauth_client_;
+  OAuth2FilterConfigSharedPtr config_;
+  Stats::Scope& scope_;
+
+  const RequirementsMap& escapedReplacements() const {
```

Comment:
we should have this in some utility, doesn't chromium_url do this? @htuch?

Reply:
ping @htuch

Reply:
Yeah, can we put this in source/common/http/utility.{cc,h} for now? We'll want to replace this this when we have GURL imported.CC @dio